File: results/0.25/ELF/elf-gemini-1.5-flash.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>

typedef struct {
    unsigned char e_ident[16];
    uint16_t e_type;
    uint16_t e_machine;
    uint32_t e_version;
    uint64_t e_entry;
    uint64_t e_phoff;
    uint64_t e_shoff;
    uint32_t e_flags;
    uint16_t e_ehsize;
    uint16_t e_phentsize;
    uint16_t e_phnum;
    uint16_t e_shentsize;
    uint16_t e_shnum;
    uint16_t e_shstrndx;
} elf_header_t;

typedef struct {
    uint32_t p_type;
    uint32_t p_flags;
    uint64_t p_offset;
    uint64_t p_vaddr;
    uint64_t p_paddr;
    uint64_t p_filesz;
    uint64_t p_memsz;
    uint64_t p_align;
} elf_program_header_t;

typedef struct {
    uint32_t sh_name;
    uint32_t sh_type;
    uint64_t sh_flags;
    uint64_t sh_addr;
    uint64_t sh_offset;
    uint64_t sh_size;
    uint32_t sh_link;
    uint32_t sh_info;
    uint64_t sh_addralign;
    uint64_t sh_entsize;
} elf_section_header_t;

HParser elf_uint16_parser() {
    return h_map(h_consume(2), (HMapFunc)HBytesToUint16);
}

HParser elf_uint32_parser() {
    return h_map(h_consume(4), (HMapFunc)HBytesToUint32);
}

HParser elf_uint64_parser() {
    return h_map(h_consume(8), (HMapFunc)HBytesToUint64);
}

HParser elf_header_parser() {
    return h_map(h_sequence(
                    h_consume(16),
                    elf_uint16_parser(),
                    elf_uint16_parser(),
                    elf_uint32_parser(),
                    elf_uint64_parser(),
                    elf_uint64_parser(),
                    elf_uint64_parser(),
                    elf_uint32_parser(),
                    elf_uint16_parser(),
                    elf_uint16_parser(),
                    elf_uint16_parser(),
                    elf_uint16_parser(),
                    elf_uint16_parser(),
                    elf_uint16_parser()),
                (HMapFunc)HBytesToElfHeader);
}


int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <elf_file>\n", argv[0]);
        return 1;
    }

    int fd = open(argv[1], O_RDONLY);
    if (fd == -1) {
        perror("open");
        return 1;
    }

    off_t fileSize = lseek(fd, 0, SEEK_END);
    lseek(fd, 0, SEEK_SET);

    void* fileBuffer = malloc(fileSize);
    if (read(fd, fileBuffer, fileSize) != fileSize) {
        perror("read");
        free(fileBuffer);
        close(fd);
        return 1;
    }
    close(fd);

    HParseResult* result = h_parse(elf_header_parser(), fileBuffer, fileSize);

    if (h_result_is_success(result)) {
        elf_header_t* header = (elf_header_t*)h_result_get_value(result);
        printf("ELF header parsed successfully!\n");
        free(header);
        h_result_destroy(result);
    } else {
        fprintf(stderr, "ELF header parsing failed: %s\n", h_result_get_error(result));
        h_result_destroy(result);
    }

    free(fileBuffer);
    return 0;
}

uint16_t HBytesToUint16(const void* bytes) {
    return *(uint16_t*)bytes;
}

uint32_t HBytesToUint32(const void* bytes) {
    return *(uint32_t*)bytes;
}

uint64_t HBytesToUint64(const void* bytes) {
    return *(uint64_t*)bytes;
}

elf_header_t* HBytesToElfHeader(const void* bytes) {
    elf_header_t* header = malloc(sizeof(elf_header_t));
    memcpy(header, bytes, sizeof(elf_header_t));
    return header;
}

File: results/0.0/ELF/elf-gemini-1.5-flash.rs
use nom::{
    be_u16, be_u32, be_u64, bytes::complete::take, combinator::map, IResult,
    number::complete::le_u32,
};
use std::env;
use std::fs::File;
use std::io::Read;

#[derive(Debug)]
struct ElfHeader {
    e_ident: [u8; 16],
    e_type: u16,
    e_machine: u16,
    e_version: u32,
    e_entry: u64,
    e_phoff: u64,
    e_shoff: u64,
    e_flags: u32,
    e_ehsize: u16,
    e_phentsize: u16,
    e_phnum: u16,
    e_shentsize: u16,
    e_shnum: u16,
    e_shstrndx: u16,
}


fn elf_header(input: &[u8]) -> IResult<&[u8], ElfHeader> {
    let (input, e_ident) = take(16usize)(input)?;
    let (input, e_type) = be_u16(input)?;
    let (input, e_machine) = be_u16(input)?;
    let (input, e_version) = be_u32(input)?;
    let (input, e_entry) = be_u64(input)?;
    let (input, e_phoff) = be_u64(input)?;
    let (input, e_shoff) = be_u64(input)?;
    let (input, e_flags) = be_u32(input)?;
    let (input, e_ehsize) = be_u16(input)?;
    let (input, e_phentsize) = be_u16(input)?;
    let (input, e_phnum) = be_u16(input)?;
    let (input, e_shentsize) = be_u16(input)?;
    let (input, e_shnum) = be_u16(input)?;
    let (input, e_shstrndx) = be_u16(input)?;

    Ok((
        input,
        ElfHeader {
            e_ident,
            e_type,
            e_machine,
            e_version,
            e_entry,
            e_phoff,
            e_shoff,
            e_flags,
            e_ehsize,
            e_phentsize,
            e_phnum,
            e_shentsize,
            e_shnum,
            e_shstrndx,
        },
    ))
}


fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        println!("Usage: elf_parser <elf_file>");
        return;
    }

    let filename = &args[1];
    let mut file = File::open(filename).expect("Failed to open file");
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer).expect("Failed to read file");

    match elf_header(&buffer) {
        Ok((remaining, header)) => {
            println!("ELF Header: {:?}", header);
            println!("Remaining bytes: {} ", remaining.len());
        }
        Err(e) => println!("Error parsing ELF header: {:?}", e),
    }
}
